scintillometry/fourier/base.py

```diff
@@ -0,0 +1,424 @@
+    `~numpy.fft.rfft` performs a real-input transform on one dimension of the
+    input, halving that dimension's length in the output.
+
+    Currently does not support Hermitian FFTs (`~numpy.fft.hfft`, etc.).
+    """
+
+    def __call__(self, time_data=None, freq_data=None, axis=0,
+                 ortho=False, sample_rate=None):
+        """Set up FFT.
+
+        Parameters
+        ----------
+        time_data : `~numpy.ndarray` or dict
+            Dummy array with dimensions and dtype of time-domain data.  Can
+            alternatively give a dict with 'shape' and 'dtype' entries.
+        freq_dtype : str
+            dtype of frequency-domain data.
+        axis : int, optional
+            Axis to transform.  Default: 0.
+        ortho : bool, optional
+            Whether to use orthogonal normalization.  Default: `False`.
+        sample_rate : float or `~astropy.units.Quantity`
+            Sample rate.
+        """
+        # Set direction, axis and normalization.  If axis is None, set it to 0.
+        axis = operator.index(axis)
+        ortho = bool(ortho)
+
+        # Store time and frequency-domain array shapes.
+        data_format = self.get_data_format(time_data=time_data,
+                                           freq_data=freq_data, axis=axis)
+
+        NumpyFFT = self._setup_fft_class(data_format, axis, ortho, sample_rate)
+
+        return NumpyFFT
+
+    def _setup_fft_class(self, data_format, axis, ortho, sample_rate):
+        """Create FFTBase subclass that wraps `numpy.fft`."""
+
+        # Calculate _norm from ortho.
+        _norm = 'ortho' if ortho else None
+
+        if data_format['time_dtype'].kind == 'c':
+
+            def _forward_fft(self, a):
+                return np.fft.fft(a, axis=self.axis, norm=self._norm).astype(
+                    self.data_format['freq_dtype'], copy=False)
+
+            def _inverse_fft(self, A):
+                return np.fft.ifft(A, axis=self.axis, norm=self._norm).astype(
+                    self.data_format['time_dtype'], copy=False)
+
+        else:
+
+            def _forward_fft(self, a):
+                return np.fft.rfft(a, axis=self.axis, norm=self._norm).astype(
+                    self.data_format['freq_dtype'], copy=False)
+
+            # irfft needs explicit length for odd-numbered outputs.
+            def _inverse_fft(self, A):
+                return np.fft.irfft(
+                    A, n=self.data_format['time_shape'][axis], axis=self.axis,
+                    norm=self._norm).astype(self.data_format['time_dtype'],
+                                            copy=False)
+
+        def constructor(self, direction='forward'):
+            super(NumpyFFT, self).__init__(direction=direction)
+            if self.direction == 'forward':
+                self._fft = self._forward_fft
+            else:
+                self._fft = self._inverse_fft
+
+        NumpyFFT = type("NumpyFFT", (FFTBase,), {
```

Comment:
I know I recommended this, but now that I see it, I wonder why we don't just do
```
class NumpyFFT(FFTBase):
    _data_format = data_format
    ...
    if data_format['time_type'].kind == 'c':
        def __call__(self, ...):
```
That's definitely clearer.